lib/Serialization/SerializeSIL.cpp

```diff
@@ -2208,7 +2213,17 @@ void SILSerializer::writeSILVTable(const SILVTable &vt) {
   if (!ShouldSerializeAll &&
       vt.getClass()->getEffectiveAccess() < swift::AccessLevel::Public)
     return;
-  VTableList[vt.getClass()->getName().str()] = NextVTableID++;
+
+  // Use the mangled name of the class as a key to distinguish between classes
+  // which have the same name (but are in different contexts).
+  Mangle::ASTMangler mangler;
+  std::string mangledClassName = mangler.mangleNominalType(vt.getClass());
+  size_t nameLength = mangledClassName.size();
+  char *stringStorage = (char *)StringTable.Allocate(nameLength, 1);
+  std::memcpy(stringStorage, mangledClassName.data(), nameLength);
```

Comment:
Can you use ASTContext::AllocateCopy() instead? Alternatively I think StringMap from llvm is designed to solve this problem

Reply:
This is probably more efficient than StringMap, since it's bump-pointer-allocated, but StringMap is simpler and that's probably better anyway. Definitely don't put it on the ASTContext because that doesn't get freed.

Reply:
StringMap does not really help here because I cannot (easily) change the type of VTableList.

Reply:
Ah, being a MapVector does make this trickier—you can't just stick a StringMap in there. Okay.